### PR TITLE
Refactor BlockStructure service and providers

### PR DESCRIPTION
**Pull request by mavasani**
This change refactors ~~and moves~~ the BlockStructure service and providers ~~to the shared Workspace utilities layer~~. The core APIs are now based off SyntaxTree instead of Document to allow the service to be used from an analyzer context, which is required for #49197.

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/Structure/BlockStructureHelpers.cs
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/Structure/BlockStructureHelpers.cs:13]: `BlockStructureServiceWithProviders` was split to move the core structure creation APIs to `BlockStructureHelpers` helper class. All the code shown as deleted in this file is just in a separate file `BlockStructureServiceWithProviders.cs`

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/Structure/VisualBasicBlockStructureProvider.vb
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/Services/Structure/VisualBasicBlockStructureProvider.vb:13]: All deleted code just moved to a separate file.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/Structure/BlockStructureServiceWithProviders.cs
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/LanguageServices/Structure/BlockStructureServiceWithProviders.cs:16]: All existing code, except `CreateContextAsync` at the end of this file.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/LanguageServices/VisualBasicBlockStructureService.vb
Comment (mavasani) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/VisualBasic/LanguageServices/VisualBasicBlockStructureService.vb:13]: All existing code.

**Comment by CyrusNajmabadi** [comment]
> This change refactors and moves the BlockStructure service and providers to the shared Workspace utilities layer

can we move to Features level instead?  Does this need to be in workspace layer itself?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/Structure/BlockStructureOptionProvider.cs
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/Structure/BlockStructureOptionProvider.cs:26]: i'm curious what other options we need.  should we just have a BlockStructureOptions that contains the bools that we need to look at while processing?

**Comment by CyrusNajmabadi** [comment]
Just to check, but TS/JS will still work with this right?  We still have the same plugin architecture for them there?  And only the C#/VB impl now defers to helpers which can now be shared?

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by mavasani** [comment]
> Just to check, but TS/JS will still work with this right? We still have the same plugin architecture for them there? And only the C#/VB impl now defers to helpers which can now be shared?

Yes, nothing changes for any languages with this PR. Only thing this PR does is now we can potentially invoke `BlockStructureHelpers.GetStructure` from analyzer layer (whenever we convert LSIF generator to an analyzer) for C# and VB.

**Comment by mavasani** [comment]
> can we move to Features level instead? Does this need to be in workspace layer its

Yes, as we discussed offline, I'll do the following:

1. Add a new shared Features layer that gets included in Features projects and CodeStyle projects
2. Open a follow-up design issue to convert all the IDE shared projects into separate utility projects with public APIs with weak public API support. These projects will have no IVTs and will be used by Workspaces/Features and CodeStyle Analyzers/Fixes respectively. This would likely need RPS approval upfront as we will be loading bunch of additional assemblies on VS startup (WorkspaceUtilities, FeaturesUtilities, CSharpWorkspaceUtilities (and/or BasicWorkspaceUtilities) and CSharpFeaturesUtilities (and/or BasicFeaturesUtilities).

**Comment by mavasani** [comment]
@CyrusNajmabadi I will do the refactoring to move the shared functionality into a separate shared project with a follow up PR. It also gives us time to ensure everyone on the team is fine with it (have started an email thread with the IDE team to get consensus). I will like to merge this PR to unblock https://github.com/dotnet/roslyn/pull/49197

**Comment by sharwell** [comment]
Note that the block structure service is obsolete and we have a long-standing request to change to the new APIs. I forget the name of the new API so I was not able to find the tracking issue in search.

**Comment by CyrusNajmabadi** [comment]
> Note that the block structure service is obsolete and we have a long-standing request to change to the new APIs

The block structure stuff is our internal API for ths. It still remains valid (and independent of any host).  There is a request to hook this up to the new API on the editor side.  However, something blocking that on our side has been a request or us to tie information to view-oriented coordinates.  Something which we haven't wanted to do as we want our APIs to be view independent.

**Comment by mavasani** [comment]
@CyrusNajmabadi I have reverted all the file moves to shared layer - all block structure code now lies in the Features layer itself.

**Pull request merged**
